File: Backtracking_and_Search/LC_17_Letter_Combinations_of_a_Phone_Number.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.19

class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        if not digits:
            return []
        mapping = {}
        mapping['2'] = "abc"
        mapping['3'] = "def"
        mapping['4'] = "ghi"
        mapping['5'] = "jkl"
        mapping['6'] = "mno"
        mapping['7'] = "pqrs"
        mapping['8'] = "tuv"
        mapping['9'] = "wxyz"
        
        ans = []

        def dfs(digits, idx, curr, ans):
            if idx == len(digits):
                ans.append(curr[:])
                return
            for d in mapping[digits[idx]]:
                dfs(digits, idx+1, curr + d, ans)

        dfs(digits, 0, "", ans)
        return ans

File: Backtracking_and_Search/LC_17_Letter_Combinations_of_a_Phone_Number.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.19

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        if (digits.size() == 0){
            return {};
        }
        mapping.insert({'2', {'a', 'b', 'c'}});
        mapping.insert({'3', {'d', 'e', 'f'}});
        mapping.insert({'4', {'g', 'h', 'i'}});
        mapping.insert({'5', {'j', 'k', 'l'}});
        mapping.insert({'6', {'m', 'n', 'o'}});
        mapping.insert({'7', {'p', 'q', 'r', 's'}});
        mapping.insert({'8', {'t', 'u', 'v'}});
        mapping.insert({'9', {'w', 'x', 'y', 'z'}});
        vector<string> ans;
        string cur; 
        dfs(digits, 0, cur, ans);
        return ans;
    }
private:
    unordered_map<char, vector<char>> mapping;
    void dfs(string digits, int i, string cur, vector<string>& ans){
        if (cur.size() == digits.size()){
            ans.push_back(cur);
            return;
        }
        for (char c : mapping[digits[i]]){
            cur += c;
            dfs(digits, i+1, cur, ans);
            cur.pop_back();
        }
    }
};